html: Update tests for an autofocus algorithm update

tkent-google opened the pull request:
https://github.com/whatwg/html/pull/4763

tkent-google committed:
html: Update tests for an autofocus algorithm update

https://github.com/whatwg/html/pull/4763

domenic commented on html/semantics/forms/autofocus/focusable-area-in-top-document.html:
```diff
@@ -0,0 +1,25 @@
+  assert_equals(document.activeElement, iframe);
+
+  let input = document.createElement('input');
+  input.autocous = true;
```
```suggestion
  input.autofocus = true;
```

tkent-google replied:
Done.

domenic commented on html/semantics/forms/autofocus/skip-document-with-fragment.html:
```diff
@@ -0,0 +1,26 @@
+  await waitForAnimationFrame();
+  await waitForAnimationFrame();
+  assert_equals(document.activeElement, input);
+}, 'Autofocus element in a document with a URI fragment should be skipped.');
```
```suggestion
}, 'Autofocus elements in iframed documents with URL fragments should be skipped.');
```

tkent-google replied:
Done.

domenic reviewed (changes requested):
Found a typo bug that needs fixing.

Things I see in the spec but I am not sure if there are tests for:

- Sandboxing should prevent autofocus
- Cross-origin should prevent autofocus
- It should wait for script-blocking style sheets, possibly for a long time. (Best done with a slow server response for a stylesheet.)
- If topDocument's URL fragment is not empty, should not autofocus. (Can test using window.open() to create a new top-level BC)
- Even with no blocking style sheets, if you wait a long time (~100 ms?), and the focus is still on the document itself, inserting an autofocus element will cause that element to be focused after the next frame.

Also worth noting that we should test the exact position in the event loop's "update the rendering" for when autofocusing happens. I'm not sure the best way to do that with the current spec proposal, although I think your use of double-rAF instead of single-rAF is probably because of the fact that it's after rAF. But for example, if we stick with the current spec proposal, we should test that after a single rAF or a single IntersectionObserver callback, there is no autofocus, but before the second resize event, the autofocus does happen.

I don't think it's necessary to land all these tests, but given how historically non-interoperable this attribute is, the more tests we can land, the better.

tkent-google committed:
Fix a typo in focusable-area-in-top-document.html, add new test cases

tkent-google commented:
> Sandboxing should prevent autofocus

Added no-sandboxed-automatic-features.html.

> Cross-origin should prevent autofocus

We already have no-cross-origin-autofocus.sub.html.

> It should wait for script-blocking style sheets, possibly for a long time. (Best done with a slow server response for a stylesheet.)

Added spin-by-blocking-style-sheet.html.

> If topDocument's URL fragment is not empty, should not autofocus. (Can test using window.open() to create a new top-level BC)

Added new test case to skip-document-with-fragment.html.

> Even with no blocking style sheets, if you wait a long time (~100 ms?), and the focus is still on the document itself, inserting an autofocus element will cause that element to be focused after the next frame.

Added autofocus-on-stable-document.html.

> Also worth noting that we should test the exact position in the event loop's "update the rendering" for when autofocusing happens. ...

Added update-the-rendering.html.

tkent-google committed:
Remove debug code

tkent-google committed:
Avoid setTimeout()

tkent-google commented:
update-the-rendering.html is flaky with Firefox because the order of an animation frame callback and a user interaction task, in which autofocus processing is currently done, is not deterministic.  I have no idea how to fix this issue, and might remove the test.

domenic reviewed (approved):
If Firefox implemented the proposed spec, then the test would not be flaky, right? In that case I think we should keep the test, and ask the admins for an exception to allow merging this.

tkent-google commented:
> If Firefox implemented the proposed spec, then the test would not be flaky, right?

That's right.

tkent-google committed:
Update for the new commit of the corresponding PR.
- Add first-reconnected.html
- Add skip-another-top-level-browsing-context.html

tkent-google committed:
Adjust existing waitUntilStableAutofocusState() callers

foolip commented:
Should this be merged now? I can admin merge if needed.

domenic commented:
@foolip yes, please admin merge!

foolip commented:
Done!
